File: core/src/ru/myitschool/dcrawler/skills/targeting/TargetRenderer.java
package ru.myitschool.dcrawler.skills.targeting;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.skills.Skill;
import ru.myitschool.dcrawler.skills.Target;
import ru.myitschool.dcrawler.tiles.ColorTile;
import ru.myitschool.dcrawler.utils.Utils;

/**
 * Created by Voyager on 28.02.2018.
 */
public class TargetRenderer {

    private Skill skill;

    private Array<TargetDisplayer> displayers = new Array<TargetDisplayer>();

    private static final ColorTile defaultTargetTile = new ColorTile(Color.RED, 0.6f, true);
    private static final ColorTile defaultLineTile = new ColorTile(Color.CYAN, 0.6f, true);
    private static final ColorTile defaultObstructionTile = new ColorTile(Color.FIREBRICK, 1, true);

    public TargetRenderer(Skill skill) {
        this.skill = skill;
    }

    public Array<TilePos> displayTarget(int x, int y){//TODO wall checks
        Array<TilePos> array = new Array<TilePos>();
        Entity doer = skill.getDoer();

        int distanceMin = skill.getDistanceMin();
        int distanceMax = skill.getDistanceMax();
        if (Utils.isTargetInDistance(x, y, doer.getTileX(), doer.getTileY(), distanceMin, distanceMax) && !skill.hasTarget(new Target(x, y))){
            targetCreation(x, y, array);
        }
        return array;
    }

    protected Array<TilePos> targetCreation(int x, int y, Array<TilePos> array){
        boolean activate = true;
        for (TargetDisplayer displayer : displayers){
            if (activate){
                activate = displayer.targetCreation(x, y, array, getSkill());
            }
        }
        return array;
    }

    public void addDisplayer(TargetDisplayer displayer){
        displayers.add(displayer);
    }

    public void clearDisplayers(){
        displayers.clear();
    }

    public static ColorTile getDefaultTargetTile() {
        return defaultTargetTile;
    }

    public static ColorTile getDefaultLineTile() {
        return defaultLineTile;
    }

    public static ColorTile getDefaultObstructionTile() {
        return defaultObstructionTile;
    }

    public Skill getSkill() {
        return skill;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/targeting/TargetRenderer.kt
package com.wolg_vlad.dcrawler.entities.skills.targeting

import com.badlogic.gdx.graphics.Color
import com.wolg_vlad.dcrawler.entities.skills.Skill
import com.wolg_vlad.dcrawler.entities.skills.Target
import com.wolg_vlad.dcrawler.ui.tiles.ColorTile
import com.wolg_vlad.dcrawler.utils.Utils

/**
 * Created by Voyager on 28.02.2018.
 */
class TargetRenderer(val skill: Skill) {
    val displayers = mutableListOf<TargetDisplayer>()
    fun displayTarget(x: Int, y: Int): List<TilePos> { //TODO wall checks
        val array = mutableListOf<TilePos>()
        val doer = skill.doer
        val distanceMin = skill.distanceMin
        val distanceMax = skill.distanceMax
        if (Utils.isTargetInDistance(x, y, doer!!.tileX, doer.tileY, distanceMin, distanceMax) && !skill.hasTarget(Target(x, y))) {
            targetCreation(x, y, array)
        }
        return array
    }

    protected fun targetCreation(x: Int, y: Int, list: MutableList<TilePos>): List<TilePos> {
        var activate = true
        for (displayer in displayers) {
            if (activate) {
                activate = displayer.targetCreation(x, y, list, skill)
            }
        }
        return list
    }

    fun addDisplayer(displayer: TargetDisplayer) {
        displayers.add(displayer)
    }

    fun addDisplayers(displayers: Array<TargetDisplayer>) {
        displayers.forEach {
            this.displayers.add(it)
        }
    }

    companion object {
        val defaultTargetTile = ColorTile(Color.RED, 0.6f, true)
        val defaultLineTile = ColorTile(Color.CYAN, 0.6f, true)
        val defaultObstructionTile = ColorTile(Color.FIREBRICK, 1F, true)

    }

}